lib/diagnostics_channel.js

```diff
@@ -0,0 +1,143 @@
+      ArrayPrototypePush(this._subscribers, subscription);
+      return;
+    }
+
+    if (typeof subscription === 'object' &&
+        typeof subscription.onMessage === 'function') {
+      let shouldListen = true;
+
+      if (typeof subscription.shouldListen === 'function') {
+        shouldListen = subscription.shouldListen(this.name);
+      }
+
+      if (shouldListen) {
+        const fn = FunctionPrototypeBind(subscription.onMessage, subscription);
+        ArrayPrototypePush(this._subscribers, fn);
+      }
+    }
+  }
+
+  shouldPublish() {
+    return true;
+  }
+
+  publish(data) {
+    for (var i = 0; i < this._subscribers.length; i++) {
+      this._subscribers[i](this.name, data);
+    }
+  }
+}
+
+class Publisher {
+  constructor(name) {
+    this._subscribers = undefined;
+    this.ifActive = undefined;
+    this.name = name;
+  }
+
+  subscribe(subscription) {
+    ObjectSetPrototypeOf(this, ActivePublisher.prototype);
+    this.ifActive = this;
+    this._subscribers = [];
+    this.subscribe(subscription);
+  }
+
+  shouldPublish() {
+    return false;
+  }
+
+  publish() {}
+}
+
+class DiagnosticsChannel {
+  constructor() {
+    this._channels = ObjectCreate(null);
+    this._subscribers = [];
+  }
+
+  publisher(name) {
+    let channel = this._channels[name];
+    if (channel) return channel;
+
+    if (typeof name !== 'string') {
+      throw new ERR_INVALID_ARG_TYPE('publisher', ['string'], name);
+    }
+
+    channel = new Publisher(name);
+
+    // New channels should check existing subscribers
+    for (let i = 0; i < this._subscribers.length; i++) {
+      const subscriber = this._subscribers[i];
+      channel.subscribe(subscriber);
+    }
+
+    this._channels[name] = channel;
+    return channel;
+  }
+
+  subscribe(matcher, onMessage) {
+    // Namespace
+    if (typeof matcher === 'string') {
+      this.publisher(matcher).subscribe(onMessage);
+      return;
+    }
+
+    let subscriber;
+    if (typeof matcher === 'function') {
+      subscriber = matcher;
+    } else if (typeof matcher === 'object' &&
+               typeof matcher.onMessage === 'function') {
+      subscriber = matcher;
+    }
+
+    if (typeof subscriber === 'undefined') {
+      throw new ERR_INVALID_ARG_TYPE(
+        'matcher', ['string', 'function', 'Subscriber'], matcher);
+    }
+
+    if (subscriber && ArrayPrototypeIncludes(this._subscribers, subscriber)) {
+      ArrayPrototypePush(this._subscribers, subscriber);
+
+      // New subscribers should check existing channels
+      for (const key in this._channels) {
+        const channel = this._channels[key];
+        channel.subscribe(subscriber);
+      }
+    }
+  }
+
+  publish(name, data) {
+    return this.publisher(name).publish(data);
+  }
+
+  shouldPublish(name) {
+    return this.publisher(name).shouldPublish();
+  }
+}
+
+const dc = new DiagnosticsChannel();
+
+dc.DiagnosticsChannel = DiagnosticsChannel;
+dc.Publisher = Publisher;
```

Comment:
Why are these exported? Noone should construct them.

Reply:
For the tests, mainly.

Reply:
we could move the implementation to internals an create an reexporting wrapper here similar as for http2.